TLS 1.3: SRV: Implement mbedtls_ssl_read_early_data()

ronald-cron-arm opened the pull request:
## Description
Fix #6341 

Note for reviewers:
Following the work done in #6621, the implementation of `mbedtls_ssl_read_early_data()` started in #6541 with the function following a logic similar as the `mbedtls_ssl_read()` one with the function triggering an handshake sequence until some early data are found and returned to the caller. Working with that guideline I realized that the resulting function could also be used easily as a companion of the mbedtls_ssl_handshake/read/write() functions. These functions all trigger an handshake sequence and if as part of this sequence early data are detected, the functions return with an appropriate return value and then the caller can call `mbedtls_ssl_read_early_data()` to read the early data. That's convenient to adapt current mbedtls_ssl_handshake/read/write() workflows to read early data as well as I experienced when updating ssl_server2 to support early data. Thus now I am wondering if we should just keep the companion behavior of `mbedtls_ssl_read_early_data()` and thus practically remove from it the call to `mbedtls_ssl_handshake()`. This would narrow its scope and thus make it easier to apprehend and use I'd say. I would appreciate opinions, thoughts about that.   

Opted for the simplest: `mbedtls_ssl_read_early_data()` does not trigger or resume the handshake (no call to `mbedtls_ssl_handshake()`). It just read early data that have been detected by a previous call to `mbedtls_ssl_handshake()`, `mbedtls_ssl_handshake_step()`, `mbedtls_ssl_handshake_read()` or `mbedtls_ssl_handshake_write()`.

## PR checklist
- [x] **changelog** not required yet
- [x] **backport** not required, new feature
- [x] **tests** provided

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -1639,6 +1639,26 @@ struct mbedtls_ssl_context {
     mbedtls_ssl_protocol_version MBEDTLS_PRIVATE(tls_version);
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+    /**
```
I'm not sure how I feel about this.

On one hand, I'm not a big fan of fields whose interpretation depend on other fields. And the fact that we're playing tricks later like giving two names to the same constant is not really making me happier on that front.

On the other hand, we want don't want to use two distinct fields for server and client because a given context is only ever going to use one of those fields, so having two would be a waste of RAM.

Now that I write it this way, this seems exactly a case for using a `union`, with two fields inside, one for server one for client.

Wdyt?

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -1639,6 +1639,26 @@ struct mbedtls_ssl_context {
+     *  information.
+     *
+     *  On server side, internal only, status of early data in the course of an
+     *  handshake. One of MBEDTLS_SSL_EARLY_DATA_STATUS_UNKNOWN,
```
Should we at least prefix the server values with `SRV` or something, to distinguish from values used client side?

mpg commented on library/ssl_tls13_server.c:
```diff
@@ -2893,6 +2893,13 @@ static int ssl_tls13_end_of_early_data_coordinate(mbedtls_ssl_context *ssl)
+         * A server receiving more than max_early_data_size bytes of 0-RTT data
+         * SHOULD terminate the connection with an "unexpected_message" alert.
+         *
+         * TODO: Add received data size check here.
```
Do we have an issue to track this?

ronald-cron-arm replied:
yes #6347

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -4739,6 +4748,11 @@ int mbedtls_ssl_get_session(const mbedtls_ssl_context *ssl,
+ *                 defined in RFC 8446 (TLS 1.3 specification), has been
+ *                 received as part of the handshake. This is server specific.
+ *                 You must call mbedtls_ssl_read_early_data() to read the
+ *                 early data before to resume the handshake.
```
Language: I think it should be "before resuming". (Multiple occurrences in this file.)

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -5065,19 +5065,31 @@ int mbedtls_ssl_close_notify(mbedtls_ssl_context *ssl);
+ *
+ * \note           This API is server specific.
+ *
+ * \note           Early data is defined in the TLS 1.3 specification, RFC 8446.
```
If Early Data has different security properties from normal data (when it comes to replay for example) this should be mentioned here.

mpg commented on library/ssl_msg.c:
```diff
@@ -5865,6 +5865,55 @@ int mbedtls_ssl_read(mbedtls_ssl_context *ssl, unsigned char *buf, size_t len)
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+    }
+
+    if ((!mbedtls_ssl_conf_is_tls13_enabled(conf)) ||
```
This seems redundant with the check below: if TLS 1.3 is not enabled, then surely it will not have been negotiated? Or am I missing something?

ronald-cron-arm replied:
Yes, belt and braces: first check the configuration then the maximum version to be negotiated or he negotiated version. I'd prefer to keep it but it you prefer I can remove it as well.

ronald-cron-arm replied:
Function re-written, not relevant anymore.

mpg reviewed (commented):
(Partial review so far.)

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -4723,6 +4748,11 @@ int mbedtls_ssl_get_session(const mbedtls_ssl_context *ssl,
  * \return         #MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED if DTLS is in use
  *                 and the client did not demonstrate reachability yet - in
  *                 this case you must stop using the context (see below).
+ * \return         #MBEDTLS_ERR_SSL_RECEIVED_EARLY_DATA if early data, as
```
For the record (as it was discussed yesterday but only but not in writing): expanding an existing function with a new return value that requires special handling is a compatibility smell and should always give us pause.

In this instance however it's fine as the new behaviour is opt-in: the application can only get this returned if it called `mbedtls_ssl_tls13_conf_early_data()` earlier. That means existing applications that used to ignore the possibility of early data will keep working as they should.

However, I think we should mention in the documentation of `mbedtls_ssl_tls13_conf_early_data()` that an application that enables early data needs to be prepared to handle `MBEDTLS_ERR_SSL_RECEIVED_EARLY_DATA` being returned by <list of functions that can now return it>. This kind of action at a distance is not obvious so it should be mentioned in the place that causes it. (If I was an existing user updating my code to support ealy data, I'd probably read the documentation of the new functions, but I most probably wouldn't go and re-read the documentation of every existing function just to check if it gained new behaviour related to early data).

ronald-cron-arm commented:
@mpg I believe I have addressed your comments. Following our discussions, I have also added documentation to highlight that mbedtls_ssl_handshake(), mbedtls_ssl_write() may now return a new error code. I have also simplified the read API.

ronald-cron-arm commented:
I have some troubles with generate_ssl_debug_helpers.py, trying to figure out the commit causing the problem. Sorry for the noise.

ronald-cron-arm commented:
CI all green but Windows-2012 not finished, I think this is ready for review.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -5101,6 +5115,21 @@ int mbedtls_ssl_close_notify(mbedtls_ssl_context *ssl);
  * \note           This API is server specific.
  *
  * \note           Early data is defined in the TLS 1.3 specification, RFC 8446.
+ *                 IMPORTANT NOTE from section 2.3 of the specification:
```
Suggestion: use `\warning` for that?

ronald-cron-arm replied:
yep seems well suited

mpg commented on library/ssl_msg.c:
```diff
@@ -5865,54 +5865,20 @@ int mbedtls_ssl_read(mbedtls_ssl_context *ssl, unsigned char *buf, size_t len)
-         * API and when the function returns 0, it only means that zero byte
-         * of early data has been received.
-         */
+    if ((ssl->state != MBEDTLS_SSL_END_OF_EARLY_DATA) ||
```
The state's name makes this a bit counter-intuitive. IIRC states are name after the next message we're expecting, but in that case we're expecting either early data or end early data. Should rename the state to reflect that? Like `MBEDTLS_SSL_EARLY_DATA_OR_END_OF_EARLY_DATA`?

Also, I'm surprised that we apparently don't need to check `ssl->early_data_state.srv` at all?

mpg replied:
> The state name is used on client side as well where MBEDTLS_SSL_EARLY_DATA_OR_END_OF_EARLY_DATA would not be suitable.

Ah, good point. Thanks for adding a comment.

mpg replied:
> Now we just need a flag to note down if the server has accepted early data or not at reception of the ClientHello. I've reworked that.

Nice, looks much simpler now. Just to make sure I understand: we don't need to check this flag here because the only way `ssl->state` can be equal to `MBEDTLS_SSL_END_OF_EARLY_DATA` is if the server has accepted early data?

Also, does `ssl->in_offt != NULL` guarantee that the server has just received early data? If that's the case that's really not obvious to me - I though `in_offt` was also used for other purposes (for example multiple handshake messages in a single record or things like that).

To put my question in context, what I'm worried about is the following scenario:
1. Application calls `ssl_handhsake()` (or even `handshake_step()`).
2. `ssl_handshake()` returns something _other_ than `RECEIVED_EARLY_DATA`.
3. Application call `mbedtls_ssl_read_early_data()` anyway because misusing the API is what users do.
4. Do (a) we detect the misuse and cleanly return an error, or (b) silently output garbage?

Right now it's not clear to me that we're doing (a) not (b) - at least not in all cases. Wdyt?

ronald-cron-arm replied:
> Nice, looks much simpler now. Just to make sure I understand: we don't need to check this flag here because the only way `ssl->state` can be equal to `MBEDTLS_SSL_END_OF_EARLY_DATA` is if the server has accepted early data?

Yes that's right, see lines 2867 to 2876. That's the only code on server side to move to the `MBEDTLS_SSL_END_OF_EARLY_DATA` state.

ronald-cron-arm replied:
> I though `in_offt` was also used for other purposes (for example multiple handshake messages in a single record or things like that).

It seems to me that `in_offt` is dedicated to application data.
Declaration in `ssl.h`:
```C
unsigned char *MBEDTLS_PRIVATE(in_offt);     /*!< read offset in application data  */
```
and looking its usage in ssl_msg.c it seems to be indeed the case.

mpg replied:
Ah right, it seems my memories were blurry and I was mixing up different things in my mind (like `in_left`, `next_record_offset` etc.). Indeed this is only for application data, so it being non-NULL guarantees that we indeed received early/application data. Thanks for clarifying!

mpg reviewed (commented):
Thanks for addressing my feedback! Looks almost all good to me, just one question and a few suggestions.

The question (about not checking `ssl->early_data_state.srv`) is the only reason I'm not approving now. The suggestions are minor and could be left to a follow-up in the interest of avoiding another round-trip.

Note about my review: at this point I'm not familiar enough with the spec and our code to be confident we're only accepting early data when we should (for example, I had completely missed the thing about HRR). So, I'm mostly assuming that negotiation of EarlyData is correct as it has been reviewed before, and also we'll have an opportunity to revisit that when adding negative testing.

ronald-cron-arm commented:
> The question (about not checking `ssl->early_data_state.srv`) is the only reason I'm not approving now. 
I have removed it.

> Note about my review: at this point I'm not familiar enough with the spec and our code to be confident we're only accepting early data when we should (for example, I had completely missed the thing about HRR). So, I'm mostly assuming that negotiation of EarlyData is correct as it has been reviewed before, and also we'll have an opportunity to revisit that when adding negative testing.

I think that's kind of normal you did not spot the need for the change relative to the reception of an HRR while reviewing this PR. The change does not fit well in this PR (I probably should have done it in another PR, sorry about that). It is indeed something we missed when working on and reviewing `ssl_tls13_update_early_data_status()` (now ssl_tls13_is_early_data_accepted()). PR 8727 adds a test where an HRR is received and thus early data rejected.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -2074,7 +2074,8 @@ void mbedtls_ssl_conf_authmode(mbedtls_ssl_config *conf, int authmode);
+ *  been received. To read the early data, call mbedtls_ssl_read_early_data()
+ *  before calling the original function again.
  *
  * \warning This interface is experimental and may change without notice.
```
Btw is that still the case? Are we planning to make this API stable before the release and if so are we tracking it?

ronald-cron-arm replied:
Not completely sure yet, we have #8772 to track it.

mpg reviewed (commented):
Thanks for addressing my feedback - and more, I really like the simplification of state handling on the server.

I still have mostly the same question though, which I've tried to clarify. To be clear: I'm not saying there's something wrong, just that I can't convince myself I know what the code is doing, at this point.

mpg reviewed (approved):
Thanks for answering my question. Looks all good to me now.

tom-cosgrove-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1644,6 +1654,18 @@ struct mbedtls_ssl_context {
+#if defined(MBEDTLS_SSL_EARLY_DATA) && defined(MBEDTLS_SSL_CLI_C)
+    /**
+     *  Status of the negotiation of the use of early data.
+     *  See the documentation of mbedtls_ssl_get_early_data_status() for more
```
Err, I can see the documentation for `mbedtls_ssl_get_early_data_status()` but I can't see the implementation - what am I missing?

ronald-cron-arm replied:
It is not implemented yet, like we have the documentation of mbedtls_ssl_write_early_data() but it is not implemented yet. It is a client specific API.

tom-cosgrove-arm reviewed (approved):
LGTM
